### ISSUE TITLE
Tracing for Lumen?

### ISSUE THREAD

**Issue by giggsey** [getsentry/sentry-laravel#410]
Hi,

Does the Tracing support Lumen?

Registering that extra Service Provider doesn't seem to work.

**Comment by addeeandra** [comment]
Had the same issue, after that I added `Tracing\Middleware::class` in `bootstrap/app.php`:

```php
$app->middleware([
    \Sentry\Laravel\Tracing\Middleware::class,
    \Fideloper\Proxy\TrustProxies::class,
    ...
]);
```

It's seems not resolving :
`Call to undefined method Laravel\Lumen\Application::booted() in Middleware.php line 104`

Then I checked the Middleware at line 104, I think this lines of code cause the issue :

```php
if (!$this->addBootTimeSpans()) {
    // @TODO: We might want to move this together with the `RouteMatches` listener to some central place and or do this from the `EventHandler`
    app()->booted(function () use ($request, $fallbackTime): void {
        $spanContextStart = new SpanContext();
        $spanContextStart->setOp('app.bootstrap');
        $spanContextStart->setStartTimestamp(defined('LARAVEL_START') ? LARAVEL_START : $request->server('REQUEST_TIME_FLOAT', $fallbackTime));
        $spanContextStart->setEndTimestamp(microtime(true));
        $this->transaction->startChild($spanContextStart);

        $appContextStart = new SpanContext();
        $appContextStart->setOp('app.handle');
        $appContextStart->setStartTimestamp(microtime(true));
        $this->appSpan = $this->transaction->startChild($appContextStart);

        SentrySdk::getCurrentHub()->setSpan($this->appSpan);
    });
}
```

**Comment by addeeandra** [comment]
Well, I did some work around.

First, by defining `LARAVEL_START`, `SENTRY_AUTOLOAD`, and `SENTRY_BOOSTRAP` somewhere at my code i.e. `define(LARAVEL_START, microtime(true))` in `bootstrap/app.php`, it will make the `$this->addBootTimeSpans()` returning true and it will just pass by the `app()->booted($closure)` statement.

It will create new SpanContext from it :

```php
// Sentry\Laravel\Tracing\Middleware.php:121

private function addBootTimeSpans(): bool
{
    if (!defined('LARAVEL_START') || !LARAVEL_START) {
        return false;
    }

    if (!defined('SENTRY_AUTOLOAD') || !SENTRY_AUTOLOAD) {
        return false;
    }

    if (!defined('SENTRY_BOOTSTRAP') || !SENTRY_BOOTSTRAP) {
        return false;
    }

    $spanContextStart = new SpanContext();
    $spanContextStart->setOp('autoload');
    $spanContextStart->setStartTimestamp(LARAVEL_START);
    $spanContextStart->setEndTimestamp(SENTRY_AUTOLOAD);
    $this->transaction->startChild($spanContextStart);

    $spanContextStart = new SpanContext();
    $spanContextStart->setOp('bootstrap');
    $spanContextStart->setStartTimestamp(SENTRY_AUTOLOAD);
    $spanContextStart->setEndTimestamp(SENTRY_BOOTSTRAP);
    $this->transaction->startChild($spanContextStart);

    return true;
}
```

---

From the documentation [Lumen - Terminable Middleware](https://lumen.laravel.com/docs/8.x/middleware#terminable-middleware) :

> When calling the terminate method on your middleware, Lumen will resolve a fresh instance of the middleware from the service container. If you would like to use the same middleware instance when the handle and terminate methods are called, register the middleware with the container using the container's singleton method.

The `terminate($request, $response)` function within `Tracing\Middleware::class` uses `$this->transaction` which is initialized in `handle()` -- `$this->transaction` will always be null, that's why **$this->transaction->finish()** never been executed and tracing will not sent to Sentry.

So I tried to extends `\Tracing\Middleware` on a new class which I called it `SentryLumenMiddleware`, it looks like this :

```php
class SentryLumenMiddleware extends Middleware
{
   ...
    public function terminate($request, $response): void
    {
        if (is_null($this->transaction) && app()->bound('sentry')) {
            /** @var Hub $sentry */
            $sentry = app('sentry');
            $this->transaction = $sentry->getTransaction();
        }
        parent::terminate($request, $response);
    }
}
```

---

By default, `\Tracing\Middleware::class` will be registered within `\Tracing\ServiceProvider::class`, but it's not possible for Lumen. Take a look at this :

```php
// Sentry\Laravel\Tracing\ServiceProvider.php:14

public function boot(): void
{
    if ($this->hasDsnSet()) {
        $this->bindEvents($this->app);

        $this->bindViewEngine();

        if ($this->app->bound(HttpKernelInterface::class)) { // <-- THIS ONE
            /** @var \Illuminate\Contracts\Http\Kernel $httpKernel */
            $httpKernel = $this->app->make(HttpKernelInterface::class);

            $httpKernel->prependMiddleware(Middleware::class);
        }
    }
}
```

Laravel has `app(HttpKernelInterface::class)`, but not in Lumen :

[lumen/tree/8.x/app/Http](https://github.com/laravel/lumen/tree/8.x/app/Http) vs [laravel/tree/8.x/app/Http](https://github.com/laravel/laravel/tree/8.x/app/Http)

So the `SentryLumenMiddleware` is safe to register within `bootstrap/app.php` for now.

---

In the end, here is how my `bootstrap/app.php` looks like :

```php
...
$app->middleware([
    \App\Http\Middleware\SentryLumenMiddleware::class,
    ...
]);

$app->register(\Sentry\Laravel\Tracing\ServiceProvider::class);
$app->register(\Sentry\Laravel\ServiceProvider::class);
...
```

It's done, and might be good for 'now' :D

Please, correct and let me know if I missed something ..

**Comment by stayallive** [comment]
Hi all, thanks for bringing this to our attention and doing some investigative work.

TL;DR: https://github.com/getsentry/sentry-laravel/pull/416

Please test out that PR to see if that does "fix" the Lumen issues, you still need to register both service providers yourself because Lumen has no auto discovery:

```php
$app->register(Sentry\Laravel\ServiceProvider::class);
$app->register(Sentry\Laravel\Tracing\ServiceProvider::class);
```

But I've fixed the issues where we assume we are in Laravel and where possible added Lumen specific code.

The middleware is already registered as a singleton by the service provider so the state issue with the missing transaction I couldn't reproduce...

Let me know if the PR works, if so we can release that and add some documentation on it!

**Comment by giggsey** [comment]
Bit delayed in getting back to this, but just tried the latest version registering the tracing service provider, and it appears to be working fine.

Closing.

**Comment by stayallive** [comment]
Thanks for the feedback, glad to hear!
